src/trio/_repl.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import annotations
+
+import ast
+import contextlib
+import inspect
+import sys
+import types
+import warnings
+from code import InteractiveConsole
+
+import trio
+import trio.lowlevel
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import BaseExceptionGroup
+
+
+class TrioInteractiveConsole(InteractiveConsole):
+    # code.InteractiveInterpreter defines locals as Mapping[str, Any]
+    # but when we pass this to FunctionType it expects a dict. So
+    # we make the type more specific on our subclass
+    locals: dict[str, object]
+
+    def __init__(self, repl_locals: dict[str, object] | None = None):
+        super().__init__(locals=repl_locals)
+        self.compile.compiler.flags |= ast.PyCF_ALLOW_TOP_LEVEL_AWAIT
+
+    def runcode(self, code: types.CodeType) -> None:
+        async def _runcode_in_trio() -> BaseException | None:
+            func = types.FunctionType(code, self.locals)
+            try:
+                coro = func()
+            except BaseException as e:
+                return e
+
+            if inspect.iscoroutine(coro):
+                try:
+                    await coro
+                except BaseException as e:
+                    return e
+            return None
+
+        maybe_exc_or_excgroup = trio.from_thread.run(_runcode_in_trio)
+
+        if maybe_exc_or_excgroup is not None:
+            # maybe_exc_or_excgroup is an exception, or an exception group.
+            # If it is SystemExit or if the exception group contains
+            # a SystemExit, quit the repl. Otherwise, print the traceback.
+            if isinstance(maybe_exc_or_excgroup, SystemExit):
+                raise SystemExit()
+            elif isinstance(maybe_exc_or_excgroup, BaseExceptionGroup):
+                for exc in maybe_exc_or_excgroup.exceptions:
+                    if isinstance(exc, SystemExit):
+                        raise SystemExit
```

Comment:
I think this will need to use `ExceptionGroup.subgroup()`, since it's possible the `SystemExit` can be nested arbitrarily deep in exception groups. It'd also be good to raise the original, not create a fresh exception - you can pass an exit code or message to `sys.exit()`, which is stored in `SystemExit` and used as the process return code.

Reply:
Oh, I did not realise exception groups could nest. I'll have a look.